pkg/apis/core/v1beta1/validation/validation_test.go

```diff
@@ -332,6 +336,278 @@ func apiResource(apiResource *metav1.APIResource) *v1beta1.APIResource {
 	}
 }
 
+func TestValidateKubeFedCluster(t *testing.T) {
+	// Validate single success case for spec and status to ensure validation
+	// functions are wired correctly.
+	statusSubResource := []bool{true, false}
+	validKFC := testcommon.ValidKubeFedCluster()
+	for _, status := range statusSubResource {
+		if errs := ValidateKubeFedCluster(validKFC, status); len(errs) != 0 {
+			t.Errorf("expected success: %v", errs)
+		}
+	}
+
+	// Validate single error case for spec and status to ensure validation
+	// functions are wired correctly.
+	type KFCAndStatusSubResource struct {
+		kfc    *v1beta1.KubeFedCluster
+		status bool
+	}
+	errorCases := map[string]KFCAndStatusSubResource{}
+
+	invalidKFCSpec := testcommon.ValidKubeFedCluster()
+	invalidKFCSpec.Spec.APIEndpoint = ""
+	errorCases["apiEndpoint: Required value"] = KFCAndStatusSubResource{
+		invalidKFCSpec,
+		false,
+	}
+
+	invalidKFCStatus := testcommon.ValidKubeFedCluster()
+	invalidKFCStatus.Status.Conditions[1].Type = ""
+	errorCases["conditions[1].type: Required value"] = KFCAndStatusSubResource{
+		invalidKFCStatus,
+		true,
+	}
+
+	for k, v := range errorCases {
+		errs := ValidateKubeFedCluster(v.kfc, v.status)
+		if len(errs) == 0 {
+			t.Errorf("[%s] expected failure", k)
+		} else if !strings.Contains(errs[0].Error(), k) {
+			t.Errorf("unexpected error: %q, expected: %q", errs[0].Error(), k)
+		}
+	}
+}
+
+func TestValidateAPIEndpoint(t *testing.T) {
+	successProtocolSchemes := []string{
+		"",
+		"http://",
+		"https://",
+	}
+	successAPIEndpoints := []string{
+		"example.com",
+		"my.example.com",
+		"192.0.2.219",
+		"[2001:db8:25a4:8d2::1]",
+	}
+	successPortNums := []string{
+		"",
+		":1",
+		":80",
+		":8080",
+		":65535",
+	}
+	successURLPath := []string{
+		"",
+		"/",
+		"/path/to/endpoint",
+	}
+
+	for _, scheme := range successProtocolSchemes {
+		for _, endpt := range successAPIEndpoints {
+			for _, port := range successPortNums {
+				for _, path := range successURLPath {
+					if scheme == "" && path == "/path/to/endpoint" {
```

Comment:
Is an empty scheme and `/` valid?

Reply:
According to `DefaultServerURL` it is. I assume it's because adding `/` still points to the root of the http server so it's interpreted the same as if it's omitted.